src/hotspot/share/utilities/waitBarrier_generic.cpp

```diff
@@ -29,66 +30,186 @@
+
+  // Try to swing cell to armed. This should always succeed after the check above.
+  int ps = Atomic::cmpxchg(&_state, -1, 1);
+  if (ps != -1) {
+    fatal("Mid arm: Cannot arm the wait barrier. State: %d", ps);
   }
-  return w;
+
+  assert(Atomic::load(&_state) > 0, "Post arm: should be armed");
 }
 
-void GenericWaitBarrier::disarm() {
-  assert(_barrier_tag != 0, "Not armed");
-  _barrier_tag = 0;
-  // Loads of _barrier_threads/_waiters must not float above disarm store and
-  // disarm store must not sink below.
-  OrderAccess::fence();
-  int left;
+int GenericWaitBarrier::Cell::wake_if_needed(int max) {
+  int wakeups = 0;
+  while (true) {
+    int cur = Atomic::load_acquire(&_outstanding_wakeups);
+    if (cur == 0) {
+      // All done, no more waiters.
+      return 0;
+    }
+    assert(cur > 0, "Sanity");
+
+    int prev = Atomic::cmpxchg(&_outstanding_wakeups, cur, cur - 1);
+    if (prev != cur) {
+      // Contention, return to caller for early return or backoff.
+      return prev;
+    }
+
+    // Signal!
+    _sem.signal();
+
+    if (wakeups++ > max) {
+      // Over the wakeup limit, break out.
+      return prev;
+    }
+  }
+}
+
+void GenericWaitBarrier::Cell::disarm() {
   SpinYield sp;
-  do {
-    left = GenericWaitBarrier::wake_if_needed();
-    if (left == 0 && _barrier_threads > 0) {
-      // There is no thread to wake but we still have barrier threads.
-      sp.wait();
+  while (true) {
+    int s = Atomic::load_acquire(&_state);
+    assert(s > 0, "Mid disarm: Should be armed. State: %d", s);
+    if (Atomic::cmpxchg(&_state, s, -s) == s) {
+      // Successfully disarmed. Wake up waiters, if we have at least one.
+      // Allow other threads to assist with wakeups, if possible.
+      int waiters = s - 1;
+      if (waiters > 0) {
+        Atomic::release_store(&_outstanding_wakeups, waiters);
+        while (wake_if_needed(INT_MAX) > 0) {
+          sp.wait();
+        }
+      }
+      break;
     }
-    // We must loop here until there are no waiters or potential waiters.
-  } while (left > 0 || _barrier_threads > 0);
-  // API specifies disarm() must provide a trailing fence.
-  OrderAccess::fence();
+    sp.wait();
+  }
+
+  assert(Atomic::load(&_outstanding_wakeups) == 0, "Post disarm: Should not have outstanding wakeups");
+  assert(Atomic::load(&_state) < 0, "Post disarm: Should be disarmed");
 }
 
-void GenericWaitBarrier::wait(int barrier_tag) {
-  assert(barrier_tag != 0, "Trying to wait on disarmed value");
-  if (barrier_tag != _barrier_tag) {
-    // API specifies wait() must provide a trailing fence.
-    OrderAccess::fence();
-    return;
+void GenericWaitBarrier::Cell::wait() {
+  // Try to register ourselves as pending waiter. If we got disarmed while
+  // trying to register, return immediately.
+  {
+    int s;
+    do {
+      s = Atomic::load_acquire(&_state);
+      if (s < 0) {
+        return;
+      }
+      assert(s > 0, "Before wait: Should be armed. State: %d", s);
```

Comment:
If we are context switched here until this cell is re-used ?
Hence we have the wrong barrier tag ?

For safepoint this can't happen since the safepoint id for this safepoint safe thread will be wrong.
Thus we can't re-use cells until this thread returns and change his safepoint id.

But it seems like this is what saves us, so if there was another use-case it could happen, no?

Reply:
I thinking adding the barrier tag into the _state(64), so it's a 32-bit tag, and as now sign bit for armed/disarmed plus the 31-bit counter.

Reply:
Aw. Yes, there is a race condition here. It is plausible that waiter here can be stuck at previous barrier tag. I thought waiting for all threads to leave the cell on `arm()` saves us here, but it does not, because we might not know at `arm()` that we have a waiter waiting to enter.

Initially I thought to include `barrier_tag` into `_state`, so we can check for it before entering in this code, but somehow convinced myself it was not required. This counter-example directly shows that we want to encode `barrier_tag` into `_state` as well. Let me do that.

Reply:
I pushed the new revision that incorporates review feedback and encodes `barrier_tag` into `_state`. It passes light testing. I am scheduling more thorough testing runs today. Feel free to take a look meanwhile!